### PR TITLE
Add mlem get started

### PR DESCRIPTION
**Pull request by mike0sv**

### CODE REVIEW

**Comment by mike0sv** [comment]
Converting this to draft since actual docs are not live on mlem.ai, but any feedback is still welcomed

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:100]: why is it called "stop tracking data"?

Comment (shcheklein) [reply example-mlem-get-started/generate.sh:100]: do we need this extra commit?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:106]: that's not very common for DVC to track each independently (esp dataset) - is it intended?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:142]: you don't need this anymore I believe

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:126]: please, let's try to keep it formatted <= 80

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:103]: do we need to use `--local` here and add `-d` http storage on top so that model is accesible?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:37]: change this?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:23]: `--use-deprecated=legacy-resolver` why is it needed - out of curiosity? is it some known bug /problem?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:80]: in this case it can be as simple as "Add data" ? (no need to use "stage")

Comment (shcheklein) [reply example-mlem-get-started/generate.sh:80]: ah, I see, it actually prepares data ... seems a bit too complicated ... should we just pre-bundle those files with code?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:75]: If you don it similar to DVC get-started you need to bundle this code and share from S3 so that people could curl or wget it

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:36]: resolve this, not sure this is needed at all in this case - it's simpler than the DVC's project

**Review by shcheklein — CHANGES_REQUESTED** [review]
Thanks a lot for codifying this 🙏 

I put some comments here and there to address, but also:

- should we create a branch in the regular example-get-started that converts models into mlem and uses gto + deploys it. That would be cool.
- in this example - should we have something about client, deployment, gto? Just saving model might be not very interesting (not clear why would I want this at all)

File: example-mlem-get-started/generate.sh
Comment (aguschin) [reply example-mlem-get-started/generate.sh:23]: Yes, it's a known issue: https://github.com/iterative/mlem/issues/140

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:106]: there are .mlem files in this dir. Can we do something like `echo *.mlem > .dvcignore && dvc add data`?

Comment (mike0sv) [reply example-mlem-get-started/generate.sh:106]: the correct pattern was found and it is `/**/?*.mlem`. Btw, does gitignore support ! negation?

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:80]: we want to show how to save data also. So reading a file, and then re-saving it to mlem is even more compicated

**Comment by mike0sv** [comment]
1. Yes, but this could wait until MLEM release I guess
2. I am planning to add something like that later, this was for first two sections of get-started https://github.com/iterative/mlem.ai/pull/43/files#diff-1cd29580767fad4637f461df8460373ba6ec523dc598d7d3550657ed56cab101

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:23]: It's faster and more importantly works (new resolver breaks until we resolve https://github.com/iterative/mlem/issues/140)

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:106]: this does not seem to work, since we cant `dvc add .mlem/dataset/` since we lose `.mlem` files, even though they ignored. So going with `dvc add .mlem/dataset/*.csv`. Would be great to have some way to do "add everything except dvc-ignored"

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:100]: but I don't think we need an extra commit ... could you clarify `this one is suggested by DVC itself`?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh]: > Btw, does dvcignore support ! negation?

yes, it does support it

it should be working, otherwise it's a bug to my mind

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:23]: yep, but does it mean that we will recommend the same in the Install guide and README? looks a bit weird ... (not critical, but I would try to address this before the release if possible)

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:75]: raw.github.com works as well, I don't see any particular problems with that .. I would bundle though everything as a single tar (or may github has a way to download revisions as a tar with curl, I don't know)

**Comment by shcheklein** [comment]
Thanks, @mike0sv for addressing some things. I left a bit more comments, but nothing major.

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:100]: image
without those two lines. But it works without commit, I checked. Just thought commit is necessary since DVC suggested it

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh]:
```
$ cat .dvcignore 
/**/*.mlem
!/.mlem/
(.venv) (mlem) MacBook-Pro-5:example-mlem-get-started mike0sv$ dvc check-ignore .mlem -d
.dvcignore:2:!/.mlem/   .mlem
$ cat .dvcignore 
/**/*.mlem
!/.mlem/
$ dvc check-ignore .mlem -d
.dvcignore:2:!/.mlem/   .mlem
```
I was trying to make an exception for `.mlem` dir in a root, but this is what I've got

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:23]: we'll deal with it before release

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:75]: https://github.com/iterative/example-mlem-get-started/archive/refs/tags/5-switch-to-dvc.tar.gz for example

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh]: q - why do you need an exception like this? (curious)

does the same combination work for gitignore?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:75]: it contains some extra files that are not needed initially though?

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh]: My goal is to exclude all `*.mlem` files, but not `.mlem` dir in repo root. But simple `/**/*.mlem` adds `.mlem` dir to ignore list. With git it works

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:75]: it contains whole repo, yes. But I don't see any point of maintaining them separately, since they are very small and easy and also will be available in the docs themselves

Comment (shcheklein) [reply example-mlem-get-started/generate.sh:75]: it's not about the size, it's about the workflow. In the dvc.org get started you would have a command:

```
wget code.zip
unzip code.zip
git add -a -m "add code"
```

https://dvc.org/doc/start/data-pipelines#expand-to-download-example-code

including any extra files there would confuse and ruin the flow of the document. 

so, your call here - if you want something similar to DVC then consider creating a clean code package.

File: example-mlem-get-started/generate.sh
Comment (aguschin) [inline example-mlem-get-started/generate.sh:2]:
```suggestion
# See https://mlem.ai/doc/get-started
```

File: example-mlem-get-started/generate.sh
Comment (aguschin) [reply example-mlem-get-started/generate.sh:23]: also, should MLEM be installed from git here? Not from PyPi?

File: example-mlem-get-started/generate.sh
Comment (aguschin) [inline example-mlem-get-started/generate.sh:22]: Why you need to install gitpython?

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:23]: we want git since we may want to prepare example repos before new release

File: example-mlem-get-started/generate.sh
Comment (mike0sv) [reply example-mlem-get-started/generate.sh:22]: not sure)))))

Comment (shcheklein) [reply example-mlem-get-started/generate.sh:22]: this is related to the way DVC calculates it's version in the dev env - https://dvc.org/doc/install/pre-release (add hash to the end of the version). If MLEM does the same (and it should be) then you might also need this. I would check if DVC has changed this mechanism though.

Comment (shcheklein) [reply example-mlem-get-started/generate.sh:22]: thanks, created https://github.com/iterative/example-repos-dev/pull/111 and removed it from the docs (pending redeploy)

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:75]: thinking second time - I remember now why I didn't use GH (most likely) - I wanted to keep the command as short and nice as possible `wget https://code.dvc.org/get-started/code.zip` and zip instead of tar

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:29]: as far as I know it doesn't install wheels (slow downs and / or makes installing the requirements process slow / broken)

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [reply example-mlem-get-started/generate.sh:21]: is it fixed?

File: example-mlem-get-started/code/src/train.py
Comment (shcheklein) [inline example-mlem-get-started/code/src/train.py:18]: make more meaningful description

File: example-mlem-get-started/code/src/train.py
Comment (shcheklein) [inline example-mlem-get-started/code/src/train.py:15]: should we call it `model` for simplicity?

**Comment by shcheklein** [comment]
@iterative/mlem hey folks, any progress on this?

**Comment by mike0sv** [comment]
Since it's already live I think we should go ahead and merge this

**Comment by shcheklein** [comment]
@mike0sv @aguschin let's cleanup it first please, it has `fixture-model-registry` for example now besides the `get-started` project.

File: example-mlem-get-started/code/src/train.py
Comment (aguschin) [reply example-mlem-get-started/code/src/train.py:15]: Don't think it matter that much. And we have docs in place, that rely on this repo, so let's not change this.

File: example-mlem-get-started/code/src/train.py
Comment (aguschin) [reply example-mlem-get-started/code/src/train.py:18]: Let's keep this. I don't think it's a problem and we use this in docs already.

**Review by aguschin — APPROVED** [review]
I think we can merge this now. @shcheklein?

File: example-mlem-get-started/code/.devcontainer/devcontainer.json
Comment (shcheklein) [inline example-mlem-get-started/code/.devcontainer/devcontainer.json:6]: should we also install the requirements?

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:7]: please apply the standard formatting (80 symbols, etc) ... I need to check if pre-commit is setup for this repo

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:36]: Git tags - let's make it a link to https://github.com/iterative/example-mlem-get-started/tags 

Also, it looks it's a mess now - I see a few tags that start with the same number

File: example-mlem-get-started/code/src/requirements.txt
Comment (shcheklein) [inline example-mlem-get-started/code/src/requirements.txt:4]: installing DVC like this will contradict the devcontainer that installs it globally - btw, why do we need it?

File: example-mlem-get-started/code/src/requirements.txt
Comment (shcheklein) [inline example-mlem-get-started/code/src/requirements.txt:8]: in the devcontainer we also install it, should we just install the requirements there?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:41]: dead code

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:83]: Q: so, do we use main or simple? do we use main at all?

**Comment by shcheklein** [comment]
@mike0sv @aguschin did the final review, put some minor questions / comments / bugs - PTAL

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md:29]: It works currently. Do we need to investigate something?

File: example-mlem-get-started/code/.devcontainer/devcontainer.json
Comment (aguschin) [reply example-mlem-get-started/code/.devcontainer/devcontainer.json:2]: Done!

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md:7]: Done.

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md:36]: Sorry, which ones do start with the same number?
image

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: We need DVC here because in one branch (`dvc`) we storing models with DVC. We don't cover this in GS, but we use this in UC: https://mlem.ai/doc/use-cases/dvc

Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: I need to check whether this actually breaks something or if it's ok to have.

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:8]: Removed the one from devcontainer.

File: example-mlem-get-started/generate.sh
Comment (aguschin) [reply example-mlem-get-started/generate.sh:83]: We use `main` later on L140 as a base for `dvc` branch. So `main` is the common beginning for both non-DVC and DVC cases.

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: Ok, I checked and it doesn't seem to cause any problems.

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [reply example-mlem-get-started/code/README.md:29]: no, it's fine (I would not recommend using it, and we don't, but it's simpler and it's not critical). The only thing is - please fix the link that leads to the virtualenv above

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [reply example-mlem-get-started/code/README.md:36]: yes, please disregard that comment about the number

File: example-mlem-get-started/code/src/requirements.txt
Comment (shcheklein) [reply example-mlem-get-started/code/src/requirements.txt:4]: Yep, I still think it's better in the use case (which is not actually even a use case, but a tutorial if we apply the DVC.org terminology) to mention the link to the DVC install page. People can decide how to install it. Unless mlem depends on it (but them there should be something like mlem[dvc] or something)

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:127]: now, we need to change it?

Comment (aguschin) [reply example-mlem-get-started/generate.sh:127]: Yes, doing that in https://github.com/iterative/mlem.ai/pull/131

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: Ok, it's here https://github.com/iterative/mlem.ai/pull/132
Could you please review?

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md:41]: > would be nice to explain that there are two branches I think

Done :)

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:31]: we mention it, but we don't use (instead we use the Python's built-in venv module). As we agreed it's fine, but we should remove this link then

File: example-mlem-get-started/code/src/requirements.txt
Comment (shcheklein) [reply example-mlem-get-started/code/src/requirements.txt:4]: done, I meant though that I would not include DVC as a dependency ... I would mention in the tutorial something like: `We assume that DVC is installed ... use this page to install it ...`

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:7]: That's quite unusual - downside is that people who come to the repo won't see anything in it. This is not good to my mind. What is the purpose of keeping the main like this at all if we have the corresponding tags? Can it be done this way:

- `simple` becomes `main`
- `dvc` - can it be done in way that builds on top of simple - we can make a PR from `dvc` to `main` (ex-simple) to DVC-ify the project?

File: example-mlem-get-started/generate.sh
Comment (shcheklein) [inline example-mlem-get-started/generate.sh:200]: extra space

+ let's rename `iterative` into `shcheklein` (or any other name) for example ... I found that people copy-paste and push this sometimes by mistake / too early

**Comment by shcheklein** [comment]
@aguschin thanks! we are getting there :) I still have some questions though. Please see some unresolved comments.

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: Then UC won't work for those who don't have DVC installed. Alternatively, we could ask for DVC to be installed globally and remove it from `requirements.txt`. Although I'm not too fond of this - I would try to keep all installables inside the venv we create, so the user could remove it later. Maybe DVC case is an exception here, don't know.

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md:40]: https://github.com/iterative/mlem.ai/pull/132#issuecomment-1185452401

File: example-mlem-get-started/code/README.md
Comment (aguschin) [reply example-mlem-get-started/code/README.md]: Did that and committed changes. Please see https://github.com/iterative/mlem.ai/pull/132 for PR updating documentation.

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:8]: minor: git -> Git

File: example-mlem-get-started/code/README.md
Comment (shcheklein) [inline example-mlem-get-started/code/README.md:22]: It's really bad practice to depend on `gh`, we should prefer using vanilla `git` commands as were discussing before.

File: example-mlem-get-started/code/src/requirements.txt
Comment (shcheklein) [reply example-mlem-get-started/code/src/requirements.txt:4]: do we need to update mlem -> mlem[dvc-s3, dvc-azure]?

Comment (shcheklein) [reply example-mlem-get-started/code/src/requirements.txt:4]: and remove `dvc` itself?

and even in this case not a big fan of this tbh, we should have a way to install `dvc-core-s3` or something, so that CLI is not installed

not a good practice to have CLI tools as dependencies, it smells

**Review by shcheklein — DISMISSED** [review]
Some minor comments left and I didn't try to run it.

Good progress, we are almost there.

Overall, still look too complicated for the MLEM get started. I think as we come to the updates on docs we can optimize this further.

**Comment by aguschin** [comment]
Thanks for the feedback. Fixed those @shcheklein.

**Comment by shcheklein** [comment]
@aguschin one unresolved item left - https://github.com/iterative/example-repos-dev/pull/109/files#r919031250 - up to you to address or merge! Good stuff.

File: example-mlem-get-started/code/src/requirements.txt
Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: > My point is that it's rare you would include dvc as a requirement unless you use it as a library (Python API).

Ok, IIUC, we could keep this as is, cause MLEM calls DVC API and DVC should be installed with `pip`. We already state that DVC is installed from pip. As we install all other packages without `--user`, I think we should install DVC like this as well. Do I misunderstand something?

Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: Btw, from the other hand we have `fastapi` and `docker` installed with pip, although they may be installed as MLEM dependencies 🤔 Not sure what's the right approach here, but I think we should be consistent?

Comment (aguschin) [reply example-mlem-get-started/code/src/requirements.txt:4]: @shcheklein, maybe I misunderstood you here? Do you maybe suggest to `pip install mlem dvc[azure,s3]` first and then `pip install -r requirements.txt` that have all other dependencies in it?

Comment (shcheklein) [reply example-mlem-get-started/code/src/requirements.txt:4]: > , I guess your last comment was about replacing dvc[azure,s3] with mlem[dvc,dvc-azure,dvc-s3], not about this discussion itself. Forgot to do that 😅

yes! sorry about the confusion

if fastapi / docker are used only by MLEM I would definitely install them via MLEM also!

**Comment by aguschin** [comment]
All changes addressed. I've just updated https://github.com/iterative/example-mlem-get-started

Now merging https://github.com/iterative/mlem.ai/pull/132 and https://github.com/iterative/example-repos-dev/pull/109

Thanks for your help and feedback @jorgeorpinel and @shcheklein !

**Pull request merged**
